### ISSUE TITLE
Initial connection to Honda intermittently causes bad issues and reboot.

### ISSUE THREAD

**Issue by tfabris** [tfabris/BlueGigaEmpeg#60]
Sometimes when I connect to honda with SET CONTROL RECONNECT there is a problem where it "asks me the wrong question", trying to do a PDU REGISTER NOTIFY for bad things like TRACK REACHED END. This requires me to reboot the bluetooth module.

Another problem that sometimes happens is that I get a infinite loop of the Honda re-querying me for all the track metadata, over and over and over again. This works but the system is super slow at that point.

This only occurs sometimes and I suspect it has something to do with the auto-reconnect features in the iWrap6 language. Something about the head unit getting the wrong idea about the connected device. I have tried both versions of the auto reconnect (iWrap6 has two ways of doing it) and neither solve the problem.

Ideas to try to work around (trying now):
- Try and see if doing my own software reconnect in my own loop instead of the ones built into iWrap.
- See if there is some kind of device ID or capabilities identification features that you need to set on the bluetooth module. Class of Device should be good but you really haven't tried many others. See if there are other settings.

**Comment by tfabris** [comment]
The monkey reconnect did not solve the following issues with a 5-second retry:
- Crackling audio playing at high pitch occurs sometimes after initial connect or reconnect.
- One of the two Honda touchscreens says "No Device Connected" even though the device is connected fine and the other touchscreen looks fine and works corretly.

Jury is still out on the other stuff.

**Comment by tfabris** [comment]
I have checked in a new Monkey Reconnect feature and some other changes.

Monkey Reconnect SEEMS TO (not sure) fix the issue where the Bad PDU Registration would happen and I would have to reboot the unit. 

However it MIGHT (again not sure) increase the chances for a cracking high pitched audio or silent audio after first connection. But, that one is so intermittent I'm not sure it's even related to Monkey Reconnect at all. Maybe it's just a timing bug on the I2S lines? Dunno!

Let this bake a while and see what kind of frequency of repro issues you get.

**Comment by tfabris** [comment]
No repeat of crackling high pitched audio yet. I tried those repros with a warm board. Maybe that issue is a Betz board bad trace or via issue. 

Got one situation where the connection came up playing silently. After a pause and play press on the touchscreen it started making sound again. Not sure what that means.  

Got one situation where the phone beat the empeg to connect to the head unit.  Not sure if that can be fixed at all or not. It happened with auto connect too sometimes. 

I have not yet seen a repro of the  main problem: No bad pdu registrations seen yet. So that is very promising.

**Comment by tfabris** [comment]
Got another repro of the "silent startup" connection on an "actually cold" start. This one did not resolve itself on pauses and unpauses. I don't have a log of it. Also possibly interesting: Shortly after connection when it was playing silently, when I tried to pause, after a moment it would just unpause me. Still silently though. That was the same as last time.

Idea: Add AVRCP connection in places where I'm doing A2DP connection and/or seeing an A2DP connection. Maybe?

**Comment by tfabris** [comment]
Something about my most recent changes has regressed the behavior on the Honda. This morning (with a warm box) I had tons of trouble pairing and track titles didn't show up and it even crashed my head unit. Need to fix this ASAP.

**Comment by tfabris** [comment]
I fixed the regressions but there is still a problem.

Root original problem still occurs:
- Bad PDU Registration bug occurs sometimes at initial connection and the Bluetooth module must reboot at that point. Issue is intermittent and rare, but annoying.

Situation:
- Using the SET CONTROL RECONNECT command seems to cause the issue to occur.
- Using my new Monkey Reconnect feature makes that particular bug go away.

Problem:
- The new Monkey Reconnect feature induces a new bug where sometimes all the correct A2DP channels do not connect correctly and there is "Silent Playback" into a not-fully-connected bluetooth stack. Basically you get 1 of the necessary 2 A2DP channels connected, and the AVRCP channel, but the second A2DP channel is hung up and timing out.

The problem seems to occur when I try to do a repeated CALL statement and then the CALL statement times out with an error. Sometimes that error occurs up to 10-20 seconds after the initial CALL statement and now I have several CALLs stacked up, some of which respond and some of which time out.

TO DO: rewrite monkey reconnect. See if there is a way to determine if the pairing buddy is available and ready to receive a connection before trying to do the call in the first place. Instead of hammering with bad calls, try waiting until you can actually connect to the thing and then make the call.

**Comment by tfabris** [comment]
Ah, no Monkey Reconnect is bad because on some systems it's just not working. On the Onkyo I get this:

0000909 Trying to connect now to our current pairing buddy.
0000001 --^ CALL 00:09:b0:9c:87:9b 19 A2DP
0000046 CALL 0
0003553 Trying to connect now to our current pairing buddy.
0000001 --^ CALL 00:09:b0:9c:87:9b 19 A2DP
0000044 CALL 1
0002293 SSP COMPLETE 00:09:b0:9c:87:9b HCI_ERROR_AUTH_FAIL
0000006 PAIR 00:09:b0:9c:87:9b FAIL
0001028 SSP COMPLETE 00:09:b0:9c:87:9b HCI_ERROR_LMP_PDU_NOT_ALLOWED
0000017 NO CARRIER 0 ERROR 9003 L2CAP_CONNECTION_REJ_SECURITY
0000004 --^ Sending to empeg: W
0000207 Trying to connect now to our current pairing buddy.
0000001 --^ CALL 00:09:b0:9c:87:9b 19 A2DP
0000050 CALL 0
0000740 SSP COMPLETE 00:09:b0:9c:87:9b HCI_ERROR_LMP_PDU_NOT_ALLOWED
0000019 NO CARRIER 1 ERROR 9003 L2CAP_CONNECTION_REJ_SECURITY
0001006 SSP COMPLETE 00:09:b0:9c:87:9b HCI_ERROR_LMP_PDU_NOT_ALLOWED
0000019 NO CARRIER 0 ERROR 9003 L2CAP_CONNECTION_REJ_SECURITY
0000002 --^ Sending to empeg: W

**Comment by tfabris** [comment]
Another idea of a possible better workaround for the disconnect/reconnect on the Honda. Try this.
- If you get a bad PDU registration.
- Try doing LIST to see which channel is the AVRCP channel
- Do CLOSE <avrcp channel>
- Do CALL <device BT address> 17 AVRCP

See if the act of disconnecting and reconnecting the AVRCP channel only is a successful workaround for the bad PDU registration that doesn't cause a drop and reconnect of the audio stream.

**Comment by tfabris** [comment]
Tried the bouncing of the AVRCP connection. Didn't fix the issue.

But in the process of trying it, I noticed something that may be relevant -- When the bad PDU registration bug occurs, the AVRCP connection is the first thing that connects, even before the A2DP connects. It's because the system is trying to autoconnect and gets a REJ RESOURCES message for A2DP which is timing out while simultaneously it's successfully connecting an AVRCP channel when it should be waiting for A2DP to connect first.

006703 -------------------------------------
0000000 empeg player application has started.
0000001 -------------------------------------
0000001 Detected empeg bootup sequence completion, sending pause command to player.
0000006 --^ Sending to empeg: W
0000440 Pausing player on first timestamp after boot up.
0000001 --^ Sending to empeg: W
0008657 NO CARRIER 0 ERROR 9004 L2CAP_CONNECTION_REJ_RESOURCES
0000003 --^ Sending to empeg: W
0000892 CONNECT 0 AVRCP 17
0000864 AVRCP 0 GET_CAPABILITIES 2
0000003 --^ AVRCP RSP 02FF
0000069 AVRCP 0 LIST_APPLICATION_SETTING_ATTRIBUTES
0000003 --^ AVRCP RSP
0000041 AVRCP 0 GET_CAPABILITIES 3
0000003 --^ AVRCP RSP 1 2
0007309 RING 1 48:f0:7b:57:15:20 19 A2DP
0000002 --^ A2DP STREAMING START
0000028 AVRCP 0 PDU_REGISTER_NOTIFICATION 3 PLAYBACK_STATUS_CHANGED 0
0000005 --^ AVRCP NFY INTERIM 3 1 2
0000257 CONNECT 2 A2DP 19
0000004 --^ A2DP STREAMING START
0000001 --^ CALL {0} 17 AVRCP
0000065 SYNTAX ERROR
